File: 22_专题/倍增/D - Teleporter.py
# 町が N 個ある。町 i から町 Ai に移動することを K 回繰り返す。
# 町 1 から始めた時、最終的にどの町にたどり着くか？
# N<=2e5 K<=1e18


from math import floor, log2
import sys

input = sys.stdin.readline

n, k = map(int, input().split())

nexts = [int(v) - 1 for v in input().split()]  # 町 i(1≤i≤N) のテレポーターの転送先は町 A iです


# 解法一 哈希表+查找循环节(类似于 n天后的牢房)


# !解法二 倍增
maxJ = floor(log2(k)) + 1
# doubling[k][i] : 町 i から 2^k 先の町はどこか
# 最后计算时将k二进制分解即可

# 1. 初始化一步
dp = [[0] * (n + 1) for _ in range(maxJ + 1)]
for i in range(n):
    dp[0][i] = nexts[i]

# 2. 倍增
for j in range(maxJ):  # j+1<=maxJ
    for i in range(n):
        dp[j + 1][i] = dp[j][dp[j][i]]

# 3.二进制分解计算
res = 0
for bit in range(maxJ + 1):
    if (k >> bit) & 1:
        res = dp[bit][res]

print(res + 1)

File: 20_杂题/atc競プロ/AtCoder Beginner Contest 167/D - Teleporter.ts
// 町が N 個ある。町 i から町 Ai に移動することを K 回繰り返す。
// 町 1 から始めた時、最終的にどの町にたどり着くか？
// N<=2e5 K<=1e18

import { useInput } from '../ts入力'

const { input } = useInput(
  `
  6 727202214173249351
  6 5 2 5 3 2
`
)

let [a, b] = input().split(' ')
const n = Number(a)
let k = BigInt(b)

const towns = input()
  .split(' ')
  .map(v => Number(v) - 1)

let MAXJ = 0n
while (1n << MAXJ < k) MAXJ++
const dp = Array.from({ length: Number(MAXJ) + 1 }, () => new Int32Array(n + 1).fill(0))

for (let i = 0; i < n; i++) {
  dp[0][i] = towns[i]
}

for (let j = 0; j + 1 <= MAXJ; j++) {
  for (let i = 0; i < n; i++) {
    dp[j + 1][i] = dp[j][dp[j][i]]
  }
}

let res = 0
for (let bit = 0n; bit <= MAXJ; bit += 1n) {
  if ((k >> bit) & 1n) {
    res = dp[Number(bit)][res]
  }
}

console.log(res + 1)
